xds/src/main/java/com/linecorp/centraldogma/xds/k8s/v1/XdsKubernetesEndpointFetchingService.java

```diff
@@ -0,0 +1,261 @@
+    }
+
+    boolean isStarted() {
+        return executorService != null;
+    }
+
+    synchronized void start(CommandExecutor commandExecutor) {
+        if (isStarted()) {
+            return;
+        }
+        this.commandExecutor = requireNonNull(commandExecutor, "commandExecutor");
+        final ScheduledExecutorService executorService = ExecutorServiceMetrics.monitor(
+                meterRegistry, Executors.newSingleThreadScheduledExecutor(
+                        new DefaultThreadFactory("k8s-plugin-executor", true)), "k8sPluginExecutor");
+        this.executorService = executorService;
+        executorService.execute(this::start);
+    }
+
+    synchronized void stop() {
+        kubernetesWatchers.values().forEach(map -> {
+            map.values().forEach(KubernetesEndpointGroup::closeAsync);
+            map.clear();
+        });
+        kubernetesWatchers.clear();
+        final ExecutorService executorService = this.executorService;
+        try {
+            final boolean interrupted = terminate(executorService);
+            if (interrupted) {
+                Thread.currentThread().interrupt();
+            }
+        } finally {
+            this.executorService = null;
+        }
+    }
+
+    @Override
+    protected ScheduledExecutorService executor() {
+        final ScheduledExecutorService executorService = this.executorService;
+        assert executorService != null;
+        return executorService;
+    }
+
+    @Override
+    protected String pathPattern() {
+        return K8S_WATCHERS_DIRECTORY + "**";
+    }
+
+    @Override
+    protected void handleXdsResource(String path, String contentAsText, String groupName)
+            throws InvalidProtocolBufferException {
+        final Watcher.Builder watcherBuilder = Watcher.newBuilder();
+        try {
+            JSON_MESSAGE_MARSHALLER.mergeValue(contentAsText, watcherBuilder);
+        } catch (IOException e) {
+            logger.warn("Failed to parse a Watcher at {}{}. content: {}",
+                        groupName, path, contentAsText, e);
+            return;
+        }
+        final Watcher k8sWatcher = watcherBuilder.build();
+        final KubernetesEndpointGroup kubernetesEndpointGroup = createKubernetesEndpointGroup(k8sWatcher);
+        final Map<String, KubernetesEndpointGroup> watchers =
+                kubernetesWatchers.computeIfAbsent(groupName, unused -> new HashMap<>());
+
+        final String watcherName = k8sWatcher.getName();
+        final KubernetesEndpointGroup oldWatcher = watchers.get(watcherName);
+        if (oldWatcher != null) {
+            oldWatcher.closeAsync();
+        }
+        watchers.put(watcherName, kubernetesEndpointGroup);
+        kubernetesEndpointGroup.addListener(endpoints -> {
+            if (endpoints.isEmpty()) {
+                return;
+            }
+            executor().execute(
+                    () ->  pushK8sEndpoints(kubernetesEndpointGroup, groupName, endpoints, watcherName));
```

Comment:
I don't think so because they are all executed by the `executor()`.
1 `handleXdsResource` is called by the executor
2 `onFileRemoved` is called immediately
3 Thus, the created `KubernetesEndpointGroup` is closing
4 the `pushK8sEndpoints` invoked by step 1 is called by the executor but it doesn't push a commit because this condition: https://github.com/line/centraldogma/blob/37e4002b92dfc6f365af52ce01e6119fe5ee8d16/xds/src/main/java/com/linecorp/centraldogma/xds/k8s/v1/XdsKubernetesEndpointFetchingService.java#L163

Reply:
I missed that line 😅  Thanks for the pointer 👍